lib/beautiful/colorful.lua

```diff
@@ -0,0 +1,1395 @@
+
+-- {{{ New Color functions
+
+--- Color object creation functions
+-- @section color_creation
+
+-- {{{ color.new()
+
+--- Create a new color instance
+-- @param val A CSS color compatible string, decimal color value, or a table
+-- containing 3 or 4 values.
+-- @return A new color object, or nil if invalid
+-- @usage local c = colorful.color.new({30, 0.8, 0.6})
+-- -- For more examples see the top of this page
+function color.new(val)
+    if type(val) == "string" then return color.from.string(val) end
+    if type(val) == "number" then
+        if val > 0xFFFFFF then return color.from.hex4(val) end
+        return color.from.hex3(val)
+    end
+    if type(val) == "table" then
+        -- do we have red/green/blue/alpha fields?
+        if val.red ~= nil and val.green ~= nil and val.blue ~= nil then
+            local alpha = val.alpha or 255
+            return color.from.rgba(val.red, val.green, val.blue, alpha)
+        -- do we have hue/saturation/lightness/alpha fields?
+        elseif val.hue ~= nil and val.saturation ~= nil and val.lightness ~= nil then
+            local alpha = val.alpha or 1
+            return color.from.hsla(val.hue, val.saturation, val.lightness, alpha)
+        -- do we have a table of 3 or 4 values?
+        elseif #val == 3 or #val == 4 then
+            -- if val[1] [0,360] and val[2] and val[3] [0,1] we deduce HSL
+            if (between(val[1], 0, 360) and
+                between(val[2], 0, 1)   and
+                between(val[3], 0, 1)      )
+            then
+                local alpha = val[4] and clamp(val[4], 0, 1) or 1
+                return color.from.hsla(
+                    val[1],
+                    val[2],
+                    val[3],
+                    alpha
+                )
+            end
+            if (between(val[1], 0, 255) and
+                between(val[2], 0, 255) and
+                between(val[3], 0, 255)    )
+            then
+                local alpha = val[4] and clamp(val[4], 0, 255) or 255
+                return color.from.rgba(
+                    val[1],
+                    val[2],
+                    val[3],
+                    alpha
+                )
+            end
+            -- if values aren't in range, escape
+        end
+        -- don't know the table type, escape
+    end
+    -- check for C userdata (GdkRGBA) with values
+    if type(val) == "userdata" then
+        if val.red ~= nil and val.green ~= nil and val.blue ~= nil then
+            local alpha = val.alpha or 1
+            return color.from.rgba(
+                val.red * 255,
+                val.green * 255,
+                val.blue * 255,
+                alpha * 255
+            )
+        end
+    end
+    -- not a valid type, quit
+    return nil
+end
+
+-- }}} color.new()
+
+-- {{{ String
+-- {{{ Helper tables and functions
+
+local css = {
+    integer = "%d?%d?%d",   -- 0-999 (but 0-255 for values)
+    --percent = "%d?%d?%d%%", -- 0-999% (but 0-100% for values)
```

Comment:
Why this commented line?

Reply:
I forgot to delete it after I changed to creating the strings through some extra descriptor functions.